pkg/reporter/sarif.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 Fugue, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"bytes"
+	"github.com/fugue/regula/pkg/loader"
+	"github.com/fugue/regula/pkg/version"
+	"github.com/owenrumney/go-sarif/v2/sarif"
+)
+
+func SarifReporter(o *RegulaReport) (string, error) {
+	report, err := sarif.New(sarif.Version210)
+	if err != nil {
+		return "", err
+	}
+
+	run := sarif.NewRunWithInformationURI("regula", version.Homepage)
+
+	// AddDistinctArtifact is slow, use a set instead
+	artifacts := map[string]struct{}{}
+
+	// First add all rule metadata
+	addedRules := map[string]struct{}{}
+	for _, r := range o.RuleResults {
+		if _, ok := addedRules[r.RuleID]; !ok {
+			addedRules[r.RuleID] = struct{}{}
+			rule := run.AddRule(r.RuleID)
+			if r.RuleSummary != "" {
+				rule = rule.WithShortDescription(sarif.NewMultiformatMessageString(r.RuleSummary))
+			}
+			if r.RuleDescription != "" {
+				rule = rule.WithFullDescription(sarif.NewMarkdownMultiformatMessageString(r.RuleDescription))
```

Comment:
I'm still getting caught up on SARIF, but [this validator I tried](https://sarifweb.azurewebsites.net/Validation) is complaining that there's not a `text` property for `fullDescription`. I think this situation is analogous to what's in the spec for rich text messages: http://docs.oasis-open.org/sarif/sarif/v2.0/csprd01/sarif-v2.0-csprd01.html#_Toc517435941

So we should probably include a `text` property (even if it's the same value in this iteration) so that clients w/o markdown support can still render it.